### PR TITLE
Reduce cast usage for aggregate functions

### LINKED ISSUE #2029: Incorrect SQL for cast inside an aggregate (MS SQL)
Version: 5.2.3 from NuGet

Code to reproduce:

```c#
session.Save(new TestClass { Int32Prop = Int32.MaxValue });
session.Save(new TestClass { Int32Prop = Int32.MaxValue });

session.Query<TestClass>()
    .GroupBy(i => 1)
    .Select(g => new {
        s = g.Sum(i => (long)i.Int32Prop)
    })
    .ToArray();
```

Generated SQL
```sql
select cast(sum(testclass0_.Int32Prop) as BIGINT) as col_0_0_ from TestClass testclass0_
```

fails with
```
Arithmetic overflow error converting expression to data type int.
```

Looks like `sum` and `cast` should be swapped.
This runs fine:

```sql
select sum(cast(testclass0_.Int32Prop as BIGINT)) as col_0_0_ from TestClass testclass0_
```

Test class and mapping:

```c#
public class TestClass {
    public virtual int Id { get; set; }
    public virtual int? Int32Prop { get; set; }
}

public class TestClassMap : ClassMap<TestClass> {
    public TestClassMap() {
        Table("TestClass");
        Id(i => i.Id);
        Map(i => i.Int32Prop);
    }
}
```

### PR DESCRIPTION
**Pull request by maca88**
This PR uses the dialect specific SUM function to determine whether cast is needed or not and also takes nullable types into consideration, which fixes #2029.
Note:
The cast usage was not reduced for SQLite as it needs it when `sum` expression contains a parameter.

Edit: this PR has been extended to other aggregate functions.

### CODE REVIEW

File: src/NHibernate.Test/NHSpecificTest/GH2029/Fixture.cs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+		{
+			if (source == null)
+			{
+				return 0;
+			}
+			int n = 0, count = 0;
+			while ((n = source.IndexOf(substring, n, StringComparison.InvariantCulture)) != -1)
+			{
+				n += substring.Length;
+				++count;
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH2029/Fixture.cs]: In c#, pre-increment is not better than post-increment, performance wise, while being generally considered less readable.
So better use the more usual `count++;` syntax.

File: src/NHibernate.Test/NHSpecificTest/GH2029/Fixture.cs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+			using (session.BeginTransaction())
+			using (var sqlLog = new SqlLogSpy())
+			{
+				var groups = session.Query<TestClass>()
+					   .GroupBy(i => 1)
+					   .Select(g => new
+					   {
+						   s = g.Sum(i => (long) i.NullableInt32Prop)
+					   })
+					   .ToArray();
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH2029/Fixture.cs]: Has `ToArray` be used on purpose to avoid async generation?
Since we may someday support `ToArrayAsync`, I think we should use `ToList` here for now, in order to avoid changes here if `ToArrayAsync` is added.

File: src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs [hunk 3]
```diff
@@ -263,6 +265,16 @@ protected HqlTreeNode VisitNhMax(NhMaxExpression expression)
 
 		protected HqlTreeNode VisitNhSum(NhSumExpression expression)
 		{
+			var type = expression.Type.UnwrapIfNullable();
+			var nhType = TypeFactory.GetDefaultTypeFor(type);
+			if (nhType != null && _parameters.SessionFactory.SQLFunctionRegistry.FindSQLFunction("sum")
```
Comment (hazzik) [inline src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs]: I think some of this logic is not necessary.

File: src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs [hunk 4]
```diff
@@ -477,8 +489,10 @@ protected HqlTreeNode VisitUnaryExpression(UnaryExpression expression)
 				case ExpressionType.Convert:
 				case ExpressionType.ConvertChecked:
 				case ExpressionType.TypeAs:
-					if ((expression.Operand.Type.IsPrimitive || expression.Operand.Type == typeof(Decimal)) &&
-						(expression.Type.IsPrimitive || expression.Type == typeof(Decimal)))
+					var operandType = expression.Operand.Type.UnwrapIfNullable();
```
Comment (hazzik) [inline src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs]: Same here

**Comment by maca88** [comment]
I've made a more sophisticated logic that can be used also for other aggregates like `COUNT` and `AVG`. Instead of checking the `ReturnedClass` property, the logic is now comparing `SqlTypes` which should be more correct as the cast is done on the database level. The logic was applied also for `AVG` but not for `COUNT` as I found out that the registered function "count_big" for `MsSql2000Dialect` is not used for hql queries and would not add a cast for a `long` type (`COUNT` in Sql Server [returns an `int`](https://docs.microsoft.com/en-us/sql/t-sql/functions/count-transact-sql?view=sql-server-2017#return-types)). It seems that [NH-865](https://nhibernate.jira.com/browse/NH-865) was never tested. I will try to address the count issue for Sql Server in a separate PR.

File: src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs [hunk 5]
```diff
@@ -584,5 +588,96 @@ protected HqlTreeNode VisitNewArrayExpression(NewArrayExpression expression)
...
+			}
+
+			var persister = _parameters.SessionFactory.GetEntityPersister(entityName);
+			var index = persister.EntityMetamodel.GetPropertyIndexOrNull(memberExpression.Member.Name);
+			return !index.HasValue
+				? TypeFactory.GetDefaultTypeFor(expression.Type) // Not mapped
+				: persister.EntityMetamodel.PropertyTypes[index.Value];
+		}
+
+		private string TryGetEntityName(MemberExpression memberExpression)
```
Comment (maca88) [inline src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs]: Copied from #1996.

File: src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs [hunk 6]
```diff
@@ -477,13 +485,12 @@ protected HqlTreeNode VisitUnaryExpression(UnaryExpression expression)
...
-					{
-						return _hqlTreeBuilder.Cast(VisitExpression(expression.Operand).AsExpression(), expression.Type);
-					}
-
-					return VisitExpression(expression.Operand);
+					return IsCastRequired(expression.Operand, expression.Type, out var existType)
+						? _hqlTreeBuilder.Cast(VisitExpression(expression.Operand).AsExpression(), expression.Type)
+						// Make a transparent cast when an IType exists, so that it can be used to retrieve the value from the data reader
+						: existType
+							? _hqlTreeBuilder.TransparentCast(VisitExpression(expression.Operand).AsExpression(), expression.Type)
```
Comment (maca88) [inline src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs]: I have to recheck this as `TransparentCast` does support only a few types.

File: src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs (hunk 5 above)
Comment (maca88) [reply src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs]: After doing #2079 I found out that this method does not cover all cases. I will update this method and try to add additional tests that cover the edge cases.

**Comment by maca88** [comment]
Set to WIP as I need to revisit this PR, see the comments above.

File: src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs (hunk 6 above)
Comment (maca88) [reply src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs]: Added an additional check for calling `TransparentCast `, as there is an edge case when casting to a custom registered type.

File: src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs (hunk 5 above)
Comment (maca88) [reply src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs]: This method was improved and exposed for internal usage by #1996. I am thinking to leave this PR as WIP until #1996 get merged in order to replace this method with the new one.

**Comment by maca88** [comment]
Rebased and updated `TryGetEntityName` (from #1996) and `SupportsType` (from #2079) methods.

File: src/NHibernate/Util/ExpressionsHelper.cs [hunk 7]
```diff
@@ -225,10 +213,24 @@ public static class ExpressionsHelper
 							out currentComponentType);
 						break;
 					case IAbstractComponentType componentType:
+						// Concatenate the component property path in order to be able to use EntityMetamodel.GetPropertyType to retrieve the type.
+						// As GetPropertyType supports only components, do not concatenate when dealing with collection composite elements or elements.
+						if (!currentType.IsAnyType)
```
Comment (maca88) [inline src/NHibernate/Util/ExpressionsHelper.cs]: Thanks for cleaning the code. I've removed this condition as it is no longer needed because `AnyType` also implements `IAssociationType`, so it will be handled by the above `case`.

File: src/NHibernate/Util/ExpressionsHelper.cs [hunk 8]
```diff
@@ -15,5 +27,572 @@ public static class ExpressionsHelper
...
+						currentComponentType = null;
+						break;
+				}
+			}
+			
+			// When traversed to the top of the expression, return the current tracking values
+			if (memberPaths.Count == 0)
+			{
+				memberPath = currentEntityPersister != null || currentComponentType != null ? member.Path : null;
+				mappedType = GetType(currentEntityPersister, currentType, member, sessionFactory, out _);
```
Comment (hazzik) [inline src/NHibernate/Util/ExpressionsHelper.cs]: Are you using `GetType` method in other PRs? If no, I think the last argument needs to be removed.

Comment (maca88) [reply src/NHibernate/Util/ExpressionsHelper.cs]: No, I've removed it.

**Comment by maca88** [comment]
I've updated the `TryGetMappedType` method to support polymorphic queries, coalesce and conditional expressions, which will be needed for #2079 in order to detect parameter types.

File: src/NHibernate/Dialect/Function/CastFunction.cs [hunk 9]
```diff
@@ -50,13 +50,8 @@ public SqlString Render(IList args, ISessionFactoryImplementor factory)
 				{
 					throw new QueryException("invalid NHibernate type for cast(), was:" + typeName);
 				}
+
 				sqlType = factory.Dialect.GetCastTypeName(sqlTypeCodes[0]);
-				if (sqlType == null)
-				{
-					//TODO: never reached, since GetTypeName() actually throws an exception!
```
Comment (maca88) [inline src/NHibernate/Dialect/Function/CastFunction.cs]: It doesn't make sense using the type name when the dialect doesn't support the cast. If the cast would be supported, `GetCastTypeName` should return it.

**Comment by maca88** [comment]
Rebased.
